Fix range_ratio Bug in RandomDataset

jadewang21 opened the pull request:
### Description
This PR fixes a bug in the `sample` method of the `RandomDataset` class. When `range_ratio > 1`, the calculation of `input_low` and `output_low` would result in `low >= high`, causing a `ValueError: low >= high`.

#### Issue
- In the original code, `input_low = int(input_len * range_ratio)` and `input_high = input_len + 1`.
- When `range_ratio > 1` (e.g., `range_ratio=2.0`, `input_len=1020`), `input_low=2040` becomes greater than `input_high=1021`, leading to the error.
- The same issue applies to `output_low` and `output_high`.

#### Fix
Based on @ywang96's suggestion, I updated the `sample` method with a simpler sampling logic:
- Enforce `range_ratio` to be in the range `[0, 1)` (via assertion).
- Change the sampling range to `[length * (1 - range_ratio), length * (1 + range_ratio)]`, providing a symmetric range around the input/output length.
- Add logging to record the sampling range: `logger.info("Sampling input_len from [%s, %s]", input_low, input_high)`.
- Update the help text for `--random-range-ratio` in `benchmark_serving.py` to clarify its range and behavior.

#### Use Case
In my use case, I needed to simulate the real data distribution of a test dataset (`xxyy.csv`):
- `num_prefill_tokens` (input length): median 1020, 25th percentile 396, 75th percentile 1189, range 2 to 14050.
- `num_decode_tokens` (output length): median 129, 25th percentile 85, 75th percentile 395, range 7 to 1000.

With the updated logic, setting `input_len=1020` and `range_ratio=0.5`, the `input_lens` range is `[510, 1530]`, and for `output_len=129` with `range_ratio=0.5`, the `output_lens` range is `[64.5, 193.5]`. This covers the 25th to 75th percentile range (input: 396 to 1189; output: 85 to 395), meeting my testing requirements. The effect is similar to my previous setting with `range_ratio=2.0` (range `[510, 2041]` for input and `[64, 259]` for output), but the new logic is simpler and aligns with the original design intent.

### Testing
- I tested the updated logic locally with the parameters `--random-input-len 1020 --random-output-len 129 --random-range-ratio 0.5 --num-prompts 128`.
- The `ValueError: low >= high` error no longer occurs, and the dataset sampling works as expected.
- The performance trends observed in my benchmark tests are consistent with my expectations, confirming that the new logic meets my use case requirements.


Please review and provide feedback. Thank you!

jadewang21 committed:
fix range_ratio bug

jadewang21 committed:
fix the range_ration bug

jadewang21 committed:
Fix: Add newline at end of benchmark_dataset.py

jadewang21 committed:
fix:pre commit

ywang96 commented:
@JenZhao Can you take a first pass at this PR?

JenZhao commented:
thank you for identifying the bug. 

the original logic before the dataset consolidation was  [this](https://github.com/vllm-project/vllm/pull/14036/files#diff-214ef8ae13652260e1af7072a50af7a117523a4c57f918688309452c17994dacL367)

```
def sample_random_requests(
    prefix_len: int,
    input_len: int,
    output_len: int,
    num_prompts: int,
    range_ratio: float,
    tokenizer: PreTrainedTokenizerBase,
) -> list[tuple[str, int, int]]:
    prefix_token_ids = np.random.randint(0,
                                         tokenizer.vocab_size,
                                         size=prefix_len).tolist()

    input_lens = np.random.randint(
        int(input_len * range_ratio),
        input_len + 1,
        size=num_prompts,
    )
    output_lens = np.random.randint(
        int(output_len * range_ratio),
        output_len + 1,
        size=num_prompts,
    )
    offsets = np.random.randint(0, tokenizer.vocab_size, size=num_prompts)
    input_requests = []
    for i in range(num_prompts):
        prompt = tokenizer.decode(prefix_token_ids +
                                  [(offsets[i] + i + j) % tokenizer.vocab_size
                                   for j in range(input_lens[i])])

        input_requests.append((prompt, int(prefix_len + input_lens[i]),
                               int(output_lens[i]), None))

    return input_requests
```

and the related parameters have these definition in `benchmark_serving.py`.
```
    random_group.add_argument(
        "--random-input-len",
        type=int,
        default=1024,
        help=
        "Number of input tokens per request, used only for random sampling.",
    )
    random_group.add_argument(
        "--random-output-len",
        type=int,
        default=128,
        help=
        "Number of output tokens per request, used only for random sampling.",
    )
    random_group.add_argument(
        "--random-range-ratio",
        type=float,
        default=1.0,
        help="Range of sampled ratio of input/output length, "
        "used only for random sampling.",
    )
    random_group.add_argument(
        "--random-prefix-len",
        type=int,
        default=0,
        help="Number of fixed prefix tokens before random "
        " context. The length range of context in a random "
        " request is [random-prefix-len, "
        " random-prefix-len + random-prefix-len * random-range-ratio).")
```

I think the original implementation was meant to cap the `input_lens` and `output_lens` at `random_input_len` and `random_output_len`, respectively, with `random_range_ratio` adding some randomness.

I wonder if we should just add an assertion for `random_range_ratio <= 1.0` and update the definition to make it clearer — unless you have a special request in your use case?

cc @ywang96 wdyt

jadewang21 commented:
Thank you @JenZhao for reviewing my PR and providing the background on the original code! I truly appreciate your recognition of my contribution!

In my use case, I do need to support `range_ratio > 1`. I used the parameters `--random-input-len 1020 --random-output-len 129 --random-range-ratio 2.0 --num-prompts 128` to test the behavior of `RandomDataset`. The goal was to simulate the real data distribution of a special test dataset, `xxyy.csv`, which is part of a specific testing scenario. Processing this dataset directly requires additional overhead, so I aimed to approximate its input and output distribution by setting hyperparameters to observe performance trends. Based on my statistics:

- `num_prefill_tokens` (input length): median 1020, 25th percentile 396, 75th percentile 1189, range 2 to 14050.
- `num_decode_tokens` (output length): median 129, 25th percentile 85, 75th percentile 395, range 7 to 1000.

With `range_ratio=2.0`, the range of `input_lens` is [510, 2041], and the range of `output_lens` is [64, 259]. This effectively covers the 25th to 75th percentile range (input: 396 to 1189; output: 85 to 395), and with multiple samplings, it can also approach the maximum values (input 14050, output 1000). This is crucial for testing the performance of `RandomDataset` under varying length distributions.

If we restrict `range_ratio <= 1.0`, the range of `input_lens` and `output_lens` becomes very narrow (e.g., `input_lens` range would be [1020, 1021]), which cannot cover the real distribution of `xxyy.csv`, limiting the flexibility of benchmark testing.

I understand that the original design likely intended to support `range_ratio <= 1.0`. To accommodate broader testing scenarios, I modified the `sample` method in this PR to allow `range_ratio > 1` and added safeguards to ensure `low < high`. To balance the original design intent with my use case needs, I suggest the following compromise:

1. Retain the current fix (supporting `range_ratio > 1`) so users can set a larger range as needed.
2. Add a warning in `benchmark_serving.py` to alert users that `range_ratio > 1` may deviate from the original design intent.
3. Update the help text for `--random-range-ratio` to clarify that `range_ratio > 1` is supported, but the recommended default is `<= 1.0`.

For example, in `benchmark_serving.py`, we could modify it as follows:

```python
random_group.add_argument(
    "--random-range-ratio",
    type=float,
    default=1.0,
    help="Range of sampled ratio of input/output length, "
         "used only for random sampling. Values > 1.0 are supported but may deviate "
         "from the original design intent (recommended range: [0, 1]).",
)

# Add a warning
import warnings
if random_range_ratio > 1.0:
    warnings.warn("random_range_ratio > 1.0 may deviate from the original design intent. "
                  "Consider using a value <= 1.0 for compatibility with the original behavior.")

ywang96 reviewed (commented):
Thanks for the discussion @JenZhao @jadewang21

If I understand correctly, say `length = X`, `range_radio = b`, then what this PR does is:
When `b < 1`, sample from `[X * b, X]`
When `b > 1`, sample from `[X / b, X * b]`

IMO this is unnecessarily convoluted? Can we not achieve this by change the logic so that we force `b < 1` and sample from `[X * (1-b), X * (1+b)]`?

Sampling from `[1000 * 0.5, 1000 * 2]` is really the same as sampling from `1250 ± 750`. Any (uniformly) random sampling from a range can be represented by `median ± a value`.

ywang96 commented:
If changing the original logic is a concern, I think it's fine to add something like
```python
logger.info("Sampling input_len from [%s, %s]", input_len_low, input_len_high)
```

jadewang21 committed:
Update RandomDataset sampling logic to [X * (1 - b), X * (1 + b)], enforce range_ratio < 1, and add logging

Signed-off-by: jadewang21 <jadewangcn@outlook.com>

jadewang21 commented:
Thank you @ywang96 and @JenZhao for your valuable feedback and guidance throughout this PR!

I’ve updated the PR based on @ywang96's suggestion to simplify the sampling logic. The `RandomDataset` `sample` method now:
- Enforces `range_ratio` to be in `[0, 1)` (via assertion).
- Uses the symmetric sampling range `[length * (1 - range_ratio), length * (1 + range_ratio)]`.
- Includes logging to record the sampling range: `logger.info("Sampling input_len from [%s, %s]", input_low, input_high)`.

I also updated the help text for `--random-range-ratio` in `benchmark_serving.py` to clarify its range and behavior. With the new logic, my previous setting of `range_ratio=2.0` (which resulted in `input_lens` range `[510, 2041]` and `output_lens` range `[64, 259]`) can be achieved by setting `range_ratio=0.5`, which gives `input_lens` range `[510, 1530]` and `output_lens` range `[64.5, 193.5]`. This covers the 25th to 75th percentile range of my test dataset (`xxyy.csv`), and the performance trends observed in my benchmark tests are consistent with my expectations.

I’ve tested the updated logic locally with `--random-input-len 1020 --random-output-len 129 --random-range-ratio 0.5 --num-prompts 128`, and it works as expected without any errors. The new approach is simpler and aligns with the original design intent while still meeting my use case requirements.

Does this look good to you? Are there any further adjustments you’d recommend? Thank you again for your support!

JenZhao commented on benchmarks/benchmark_serving.py:
```diff
@@ -970,9 +970,10 @@ def main(args: argparse.Namespace):
+         "a symmetric sampling range [length * (1 - range_ratio), length * (1 + range_ratio)].",
     )
     random_group.add_argument(
         "--random-prefix-len",
```
let's also update the definition of random-prefix-len

jadewang21 committed:
Fix pre-commit && Update random-range-ratio default to 0.0 and improve random-prefix-len help text

ywang96 reviewed (approved):
Thanks for addressing all the comments from the discussion!

jadewang21 commented:
Thank you @JenZhao for your feedback and suggestions!

I’ve updated the PR based on your comments:
- Changed the default value of `--random-range-ratio` to `0.0` to match the original default behavior (range `[length, length]`), as suggested.
- Updated the help text for `--random-prefix-len` to clarify its role and relationship with the sampling range: "Number of fixed prefix tokens before the random context in a request. The total input length is the sum of `random-prefix-len` and a random context length sampled from [input_len * (1 - range_ratio), input_len * (1 + range_ratio)]."

I’ve tested the updated logic with `--random-input-len 1020 --random-output-len 129 --random-range-ratio 0.5 --num-prompts 128`, and it works as expected. The default behavior with `range_ratio=0.0` also aligns with the original design intent.

Does this look good to you? Are there any other adjustments you’d recommend? Thank you again for your guidance!

JenZhao commented:
Thank you, @jadewang21, for making the change!

One final suggestion — could you also update the definitions here as well?

https://github.com/vllm-project/vllm/blob/1621b252888887cc6601c5df2dfa8841ae305d8e/benchmarks/benchmark_throughput.py#L597-L609

and the default values here to match the original defaults as well?

https://github.com/vllm-project/vllm/blob/4aed0ca6a2de8c56838fe850f8d8f8559074744a/benchmarks/benchmark_dataset.py#L291

thank you!

jadewang21 committed:
Update prefix-len and random-range-ratio definitions in benchmark_throughput.py, and set DEFAULT_RANGE_RATIO to 0.0

Signed-off-by: jadewang21 <jadewangcn@outlook.com>

jadewang21 commented:
Thank you @JenZhao for your feedback and for reviewing the changes!

I’ve updated the PR based on your final suggestion:
- In `benchmark_throughput.py`, I updated the definitions of `--prefix-len` and `--random-range-ratio` to match the ones in `benchmark_serving.py`. I also set their default values to `0` and `0.0`, respectively, to align with the original default behavior.
- In `benchmark_dataset.py`, I changed `DEFAULT_RANGE_RATIO` from `1.0` to `0.0` to match the updated default behavior.

I’ve tested the changes with `--random-input-len 1020 --random-output-len 129 --random-range-ratio 0.5 --num-prompts 128`, and everything works as expected. The default behavior with `range_ratio=0.0` now matches the original behavior (range `[length, length]`).

Does this look good to you? Thank you again for your guidance!
